src/topology/homotopy.lean

```diff
@@ -0,0 +1,112 @@
+/-
+Copyright (c) 2021 Shing Tak Lam. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Shing Tak Lam
+-/
+
+import topology.unit_interval
+import topology.algebra.ordered.proj_Icc
+import topology.continuous_function.basic
+
+/-!
+# Homotopy between functions
+
+In this file, we define a `homotopy` between two functions `f₀` and `f₁`.
+
+## Definitions
+
+* `homotopy f₀ f₁` is the type of homotopies between `f₀` and `f₁`
+* `homotopy.refl f₀` is the constant homotopy between `f₀` and `f₀`
+* `homotopy.symm f₀ f₁` is a `homotopy f₁ f₀` defined by reversing the homotopy
+* `homotopy.trans F G`, where `F : homotopy f₀ f₁`, `G : homotopy f₁ f₂` is a `homotopy f₀ f₂`
+  defined by putting the first homotopy on `[0, 1/2]` and the second on `[1/2, 1]`.
+-/
+
+noncomputable theory
+
+universes u v
+
+variables {X : Type u} {Y : Type v} [topological_space X] [topological_space Y]
+
+open_locale unit_interval
+
+/--
+The type of homotopies between two functions.
+-/
+
+structure homotopy (f₀ f₁ : X → Y) extends C(X × I, Y) :=
+(to_fun_zero : ∀ x, to_fun (x, 0) = f₀ x)
+(to_fun_one : ∀ x, to_fun (x, 1) = f₁ x)
+
+namespace homotopy
+
+section
+
+variables {f₀ f₁ : X → Y}
+
+instance : has_coe_t (homotopy f₀ f₁) C(X × I, Y) := ⟨homotopy.to_continuous_map⟩
+instance : has_coe_to_fun (homotopy f₀ f₁) := ⟨_, λ F, F.to_fun⟩
+
+@[continuity]
+protected lemma continuous (F : homotopy f₀ f₁) : continuous F := F.continuous_to_fun
+
+@[simp]
+lemma apply_zero (F : homotopy f₀ f₁) (x : X) : F (x, 0) = f₀ x := F.to_fun_zero x
+@[simp]
+lemma apply_one (F : homotopy f₀ f₁) (x : X) : F (x, 1) = f₁ x := F.to_fun_one x
+
+/--
+Currying a homotopy to give us a function `X → I → Y`.
+-/
+def curry (F : homotopy f₀ f₁) : X → I → Y := function.curry F
```

Comment:
In fact, I think if you just remove this def then `F.curry` will automatically resolve to the right thing.

Reply:
Yeah, I'll refactor to use that.

Reply:
I don't know why, but it seems like with some combination of changing this, and using `continuous_map.Icc_extend`, it broke the proof of `trans`, since `continuity` can't seem to prove `⊢ continuous (λ (x : ↥I × X), ⇑(⇑(F.extend) (2 * ↑(x.fst))) x.snd)` anymore. (I changed the order of the product in `homotopy`, but that's not the issue), whereas it used to be able to prove `⊢ continuous (λ (x : ↥I × X), F.extend (2 * ↑(x.fst)) x.snd)`.

Reply:
I've put a version of this PR using `continuous_map.curry` on [`shingtaklam1324-homotopy-2`](https://github.com/leanprover-community/mathlib/tree/shingtaklam1324-homotopy2).

Reply:
Pasting in the proof that `continuity` generates into the proof does work however, so I'll do that for now, and not worry about why `continuity` is failing.